Core/include/Acts/Utilities/TransformRange.hpp

```diff
@@ -0,0 +1,228 @@
+/// for the range. It applies the callable to the value that is dereferences.
+/// It also maintains const-ness, if instructed, by returning const references
+/// only.
+/// @tparam Callable The callable to apply to the value
+/// @tparam iterator_t The iterator type of the underlying container
+template <typename Callable, typename iterator_t, bool force_const>
+struct TransformRangeIterator {
+ private:
+  using internal_value_type =
+      typename std::iterator_traits<iterator_t>::value_type;
+
+  using raw_value_type = std::remove_reference_t<decltype(Callable::apply(
+      std::declval<internal_value_type>()))>;
+
+ public:
+  /// The underlying value type that is returned by the iterator.
+  /// If @c force_const is set to true, the iterator will only return
+  /// const-qualified references
+  using value_type =
+      std::conditional_t<force_const, std::add_const_t<raw_value_type>,
+                         raw_value_type>;
+
+  using difference_type =
+      typename std::iterator_traits<iterator_t>::difference_type;
+  using pointer = std::remove_reference_t<value_type>*;
+  using reference = value_type&;
+  using iterator_category = std::forward_iterator_tag;
+
+  /// Construct an iterator from an underlying iterator
+  explicit TransformRangeIterator(iterator_t iterator) : m_iterator(iterator) {}
+
+  /// Return a reference to the value that is transformed by the callable
+  /// @return Reference to the transformed value
+  reference operator*() { return Callable::apply(*m_iterator); }
+
+  /// Return a const-reference to the value that is transformed by the callable
+  /// @return Const-reference to the transformed value
+  reference operator*() const { return Callable::apply(*m_iterator); }
+
+  /// Advance the iterator
+  /// @return Reference to the iterator
+  TransformRangeIterator& operator++() {
+    ++m_iterator;
+    return *this;
+  }
+
+  /// Compare two iterators for equality
+  /// @param other The other iterator to compare to
+  bool operator==(const TransformRangeIterator& other) const {
+    return m_iterator == other.m_iterator;
+  }
+
+  /// Compare two iterators for inequality
+  /// @param other The other iterator to compare to
+  bool operator!=(const TransformRangeIterator& other) const {
+    return m_iterator != other.m_iterator;
+  }
+
+ private:
+  iterator_t m_iterator;
+};
+
+/// Callable that dereferences a value
+struct Dereference {
+  template <typename input_t>
+  constexpr static decltype(auto) apply(input_t&& value) {
+    return *value;
+  }
+};
+
+/// Callable that const-dereferences a value
+struct ConstDereference {
+  template <typename input_t>
+  constexpr static decltype(auto) apply(input_t&& value) {
+    return std::as_const(*value);
+  }
+};
+
+/// Callable that calls the @c get method of a value
+struct DotGet {
+  template <typename input_t>
+  constexpr static decltype(auto) apply(input_t&& value) {
+    return value.get();
```

Comment:
I don't know if it makes a difference, but maybe use ```std::forward```?

Reply:
Conclusion: we're not sure if that would do anything here.